dbms/src/Interpreters/Aggregator.cpp

```diff
@@ -665,7 +669,55 @@ void NO_INLINE Aggregator::executeImpl(
 {
     typename Method::State state(agg_process_info.key_columns, key_sizes, collators);
 
-    executeImplBatch<collect_hit_rate, only_lookup>(method, state, aggregates_pool, agg_process_info);
+    // 2MB as prefetch threshold, because normally server L2 cache is 1MB.
+    static constexpr size_t prefetch_threshold = (2 << 20);
+#ifndef NDEBUG
+    bool disable_prefetch = (method.data.getBufferSizeInBytes() < prefetch_threshold);
+    fiu_do_on(FailPoints::force_agg_prefetch, { disable_prefetch = false; });
+#else
+    const bool disable_prefetch = (method.data.getBufferSizeInBytes() < prefetch_threshold);
+#endif
+
+    if constexpr (Method::State::is_serialized_key)
+    {
+        executeImplMiniBatch<collect_hit_rate, only_lookup, false>(method, state, aggregates_pool, agg_process_info);
+    }
+    else if constexpr (Method::Data::is_string_hash_map)
+    {
+        // StringHashMap doesn't support prefetch.
+        executeImplMiniBatch<collect_hit_rate, only_lookup, false>(method, state, aggregates_pool, agg_process_info);
+    }
+    else
+    {
+        if (disable_prefetch)
+            executeImplMiniBatch<collect_hit_rate, only_lookup, false>(
+                method,
+                state,
+                aggregates_pool,
+                agg_process_info);
+        else
+            executeImplMiniBatch<collect_hit_rate, only_lookup, true>(method, state, aggregates_pool, agg_process_info);
```

Comment:
If prefetching is not enabled, cache miss should not be a problem. In this case, mini-batch may increase the penalty of calling virtual functions. I think mini-batch should be used with prefetching.